sound/soc/sof/control.c

```diff
@@ -232,28 +232,26 @@ int snd_sof_bytes_put(struct snd_kcontrol *kcontrol,
 	struct sof_abi_hdr *data = cdata->data;
 	int ret, err;
 
+	if (data->size > be->max) {
+		dev_err(sdev->dev, "error: size too big %d bytes max is %d\n",
+			data->size, be->max);
+		return -EINVAL;
+	}
+
 	ret = pm_runtime_get_sync(sdev->dev);
 	if (ret < 0) {
 		dev_err(sdev->dev, "error: bytes put failed to resume %d\n",
 			ret);
 		return ret;
 	}
 
-	if (data->size > be->max) {
-		dev_err(sdev->dev, "error: size too big %d bytes max is %d\n",
-			data->size, be->max);
-		ret = -EINVAL;
-		goto out;
-	}
-
 	/* copy from kcontrol */
 	memcpy(data, ucontrol->value.bytes.data, data->size);
```

Comment:
@bardliao I think this is what takashi was talking about in his comment. ucontrol has a max size of 512 bytes. We need to make sue we copy only min(data->size, 512)

Reply:
@ranj063 for the put() case we want to check if the data will fit in 512 bytes, but you also want to check that we don't write past what was allocated. For the get() case, we really want to provide only the size of the controls. Providing kernel data to userspace is frowned upon.

Reply:
@plbossart yes you are right about the get() case.

Reply:
@plbossart @ranj063 I am thinking should we remove snd_sof_bytes_get() andsnd_sof_bytes_put() at all since we didn't test it.

Reply:
Strange, isn't there a macro defined for 512? As such it's a ridiculously low limit. A non-symmetrical stereo 192 tap FIR with 16 bits coefficients will need 768 bytes for the payload. A beamformer for 8 microphones can need e.g. 2 kBytes. Plus we have fairly large ABI headers for binary data.

Reply:
Concerning the use of bytes get/put, is it for the tone generator? We will need control for amplitude, frequency, etc. Just beep on/off is not enough.